dask/layers.py

```diff
@@ -1239,3 +1262,341 @@ def __repr__(self):
+    output_splits : list, optional
+        List of required output partitions. This parameter is used
+        internally by Dask for high-level culling.
+    tree_node_name : str, optional
+        Name to use for intermediate tree-node tasks.
+    """
+
+    def __init__(
+        self,
+        name,
+        name_input,
+        npartitions_input,
+        concat_func,
+        tree_node_func,
+        tree_node_kwargs=None,
+        finalize_func=None,
+        finalize_kwargs=None,
+        split_every=32,
+        split_out=None,
+        output_splits=None,
+        tree_node_name=None,
+        annotations=None,
+    ):
+        super().__init__(annotations=annotations)
+        self.name = name
+        self.name_input = name_input
+        self.npartitions_input = npartitions_input
+        self.concat_func = concat_func
+        self.tree_node_func = tree_node_func
+        self.tree_node_kwargs = tree_node_kwargs or {}
+        self.finalize_func = finalize_func
+        self.finalize_kwargs = finalize_kwargs or {}
+        self.split_every = split_every
+        self.split_out = split_out
+        self.output_splits = output_splits
+        self.tree_node_name = tree_node_name or "tree_node-" + self.name
+
+        # Calculate tree withs and height
+        # (Used to get output keys without materializing)
+        parts = self.npartitions_input
+        self.widths = [parts]
+        while parts > 1:
+            parts = math.ceil(parts / self.split_every)
+            self.widths.append(parts)
+        self.height = len(self.widths)
+
+        # Set default output_splits
+        if self.output_splits is None:
+            self.output_splits = list(range(self.split_out or 1))
+
+    def _check_key(self, *task):
+        # Helper function to remove the last
+        # (split_out) element in a task key
+        # when bool(split_out) is False
+        return task if self.split_out else task[:-1]
+
+    def _define_task(self, node_list, output_node=False):
+        # Define nested concatenation and func task
+        conc = (self.concat_func, node_list)
+        if output_node and self.finalize_func:
+            func = self.finalize_func
+            kwargs = self.finalize_kwargs
+        else:
+            func = self.tree_node_func
+            kwargs = self.tree_node_kwargs
+        if kwargs:
+            return (apply, func, [conc], kwargs)
+        else:
+            return (func, conc)
+
+    def _construct_graph(self, on_scheduler=False):
+        """Construct graph for a tree reduction."""
+
+        dsk = {}
+        if not self.output_splits:
+            return dsk
+
+        # Deal with pickled user-defined functions
+        if on_scheduler:
+            self.tree_node_func = CallablePickled(self.tree_node_func)
+            self.concat_func = CallablePickled(self.concat_func)
+            if self.finalize_func:
+                self.finalize_func = CallablePickled(self.finalize_func)
```

Comment:
This should not be the responsibility of `_construct_graph`. I don't like `CallablePickled` in the first place, but at the very least this wrapping should happen in `__dask_distributed_unpack__`, so graph construction is separated from serialization concerns.

Reply:
Good call! This is obviously just following the pattern of other layers, but there is really no reason for the `on_scheduler=` to be a thing (or `deserializing=` in other places).